src/serde/parse_atom.rs

```diff
@@ -0,0 +1,126 @@
+use std::io::{Cursor, Read, Result, Seek, SeekFrom};
+
+use crate::allocator::{Allocator, NodePtr};
+
+use super::errors::{bad_encoding, internal_error};
+
+const MAX_SINGLE_BYTE: u8 = 0x7f;
+
+/// decode the length prefix for an atom. Atoms whose value fit in 7 bits
+/// don't have a length prefix, so those should be handled specially and
+/// never passed to this function.
+pub fn decode_size<R: Read>(f: &mut R, initial_b: u8) -> Result<u64> {
+    debug_assert!((initial_b & 0x80) != 0);
+    if (initial_b & 0x80) == 0 {
+        return Err(internal_error());
+    }
+
+    let mut bit_count = 0;
+    let mut bit_mask: u8 = 0x80;
+    let mut b = initial_b;
+    while b & bit_mask != 0 {
+        bit_count += 1;
+        b &= 0xff ^ bit_mask;
+        bit_mask >>= 1;
+    }
+    let mut size_blob: Vec<u8> = Vec::new();
+    size_blob.resize(bit_count, 0);
+    size_blob[0] = b;
+    if bit_count > 1 {
+        let remaining_buffer = &mut size_blob[1..];
+        f.read_exact(remaining_buffer)?;
+    }
+    // need to convert size_blob to an int
+    let mut v: u64 = 0;
+    if size_blob.len() > 6 {
+        return Err(bad_encoding());
+    }
+    for b in &size_blob {
+        v <<= 8;
+        v += *b as u64;
+    }
+    if v >= 0x400000000 {
+        return Err(bad_encoding());
+    }
+    Ok(v)
+}
+
+/// parse an atom from the stream
+/// At this point, the first byte has already been read to ensure it's
+/// not a special code like `CONS_BOX_MARKER` = 0xff, so it must be
+/// passed in too
+
+pub fn parse_atom(
```

Comment:
there's a great opportunity to add unit tests for this function now that it's been factored out. e.g. I believe the case where the input buffer is too short may not be covered well

Reply:
Added.